File: test_gpu_allreduce.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main( int argc, char** argv )
{
    MPI_Init (&argc, &argv);

    int rank, size;
    int ll=20480000;
    //int ll=2048;
    int nn=10;
    int* buff = NULL;
    size_t bytes;
    int i;
    double t1, t2;

    // Get MPI rank and size
    MPI_Comm_rank (MPI_COMM_WORLD, &rank);
    MPI_Comm_size (MPI_COMM_WORLD, &size);

    // Initialize buffer
    bytes = ll*sizeof(int);
    buff = (int*)malloc(bytes);

    for(i=0; i<ll; i++) {
        buff[i] = 1;
    }
/*
    if(rank==0)
        printf("buff = %d \n", buff[5]);
*/
    if (rank==0) {
        printf("No. of ranks: %d \n", size);
        printf("No. of bytes: %d \n", bytes);
    }

    /////////////////////////////////////////////////// CPU MPI
    // warm up
    for(i=0; i<nn; i++) {
        MPI_Allreduce(MPI_IN_PLACE, buff, ll, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
    }

    t1 = MPI_Wtime();

    for(i=0; i<nn; i++) {
        MPI_Allreduce(MPI_IN_PLACE, buff, ll, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
    }

    t2 = MPI_Wtime();

    if (rank==0) {
        printf("CPU MPI: %f \n", (t2-t1)/nn);
    }

    ////////////////////////////////////////////// GPU MPI
    for(i=0; i<ll; i++) {
        buff[i] = 1;
    }
    // Copy buff to device at start of region and back to host and end of region
    #pragma acc data copy(buff[0:ll])
    {
        // warm up
        //printf("%p\n",buff);
        // Inside this region the device data pointer will be used
        #pragma acc host_data use_device(buff)
        {
            //printf("%p\n",buff);

            for(i=0; i<nn; i++) {
                MPI_Allreduce(MPI_IN_PLACE, buff, ll, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
            }
        }

        t1 = MPI_Wtime();

        #pragma acc host_data use_device(buff)
        {
            for(i=0; i<nn; i++) {
                MPI_Allreduce(MPI_IN_PLACE, buff, ll, MPI_INT, MPI_SUM, MPI_COMM_WORLD);
            }
        }

        t2 = MPI_Wtime();
    }

    if (rank==0) {
        printf("GPU MPI: %f \n", (t2-t1)/nn);
    }


    // Check that buffer is correct
/*
    if(rank==0)
        printf("buff = %d \n", buff[5]);
*/

    // Clean up
    free(buff);

    MPI_Finalize();

    return 0;
}

File: test_gpu_allreduce.f90
program test_gpu_allreduce

   !use MPI

   implicit none
   include 'mpif.h'

   !implicit none (type,external)

   integer, parameter :: dp = selected_real_kind(15,300)
   !integer, parameter :: ll = 20480000
   integer, parameter :: ll = 2048
   integer, parameter :: nn = 10

   integer :: comm
   integer :: irank
   integer :: nrank
   integer :: ierr
   integer :: ii

   real(dp) :: t1
   real(dp) :: t2

   integer, allocatable :: buf(:)
   !integer, allocatable :: buf2(:)

   call MPI_Init(ierr)

   comm = MPI_COMM_WORLD

   call MPI_Comm_rank(comm,irank,ierr)
   call MPI_Comm_size(comm,nrank,ierr)

   allocate(buf(ll))
   !allocate(buf2(ll))

   buf(:) = 1

   if(irank == 0) then
      write(6,"(2X,A,I10)") "Number of ranks:",nrank
      write(6,"(2X,A,I10)") "Number of bytes:",ll*4
      flush(6)
   end if

   !!!!!!!!!!!!!!!!!!!!! CPU MPI !!!!!!!!!!!!!!!!!!!!!

   ! warm up

   do ii = 1,nn
      call MPI_Allreduce(MPI_IN_PLACE,buf,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
      !call MPI_Allreduce(buf,buf2,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
   end do

   ! time

   t1 = MPI_Wtime()

   do ii = 1,nn
      call MPI_Allreduce(MPI_IN_PLACE,buf,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
      !call MPI_Allreduce(buf,buf2,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
   end do

   t2 = MPI_Wtime()

   if(irank == 0) then
      write(6,"(2X,A,F10.3,A)") "CPU MPI:",(t2-t1)/nn,"s"
      flush(6)
   end if

   !!!!!!!!!!!!!!!!!!!!! GPU MPI !!!!!!!!!!!!!!!!!!!!!

   buf(:) = 1

   !$acc data copy(buf)

   ! warm up

   print *, loc(buf)
   !$acc host_data use_device(buf)
   print *, loc(buf)
   do ii = 1,nn
      call MPI_Allreduce(MPI_IN_PLACE,buf,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
      !call MPI_Allreduce(buf,buf2,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
   end do
   !$acc end host_data

   ! time

   t1 = MPI_Wtime()

   !$acc host_data use_device(buf)
   do ii = 1,nn
      call MPI_Allreduce(MPI_IN_PLACE,buf,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
      !call MPI_Allreduce(buf,buf2,ll,MPI_INTEGER,MPI_SUM,comm,ierr)
   end do
   !$acc end host_data

   t2 = MPI_Wtime()

   if(irank == 0) then
      write(6,"(2X,A,F10.3,A)") "GPU MPI:",(t2-t1)/nn,"s"
      flush(6)
   end if

   !$acc end data

   deallocate(buf)
   !deallocate(buf2)

   call MPI_Finalize(ierr)

end program
